compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -455,24 +664,77 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
     val replacement = tp.dealiasKeepAnnots.stripTypeVar
-    if param == replacement then this.checkNonCyclic()
+    if param == replacement then this.checkWellFormed()
     else
       assert(replacement.isValueTypeOrLambda)
-      var current =
-        if isRemovable(param.binder) then remove(param.binder)
-        else updateEntry(this, param, replacement)
-
-      def removeParam(ps: List[TypeParamRef]) = ps.filterConserve(param ne _)
 
-      def replaceParam(tp: Type, atPoly: TypeLambda, atIdx: Int): Type =
-        current.ensureNonCyclic(atPoly.paramRefs(atIdx), tp.substParam(param, replacement))
-
-      current.foreachParam { (p, i) =>
-        current = boundsLens.map(this, current, p, i, replaceParam(_, p, i))
-        current = lowerLens.map(this, current, p, i, removeParam)
-        current = upperLens.map(this, current, p, i, removeParam)
-      }
-      current.checkNonCyclic()
+      val replacedTypeVar = typeVarOfParam(param)
+      //println(i"replace $param with $replacement in $this")
+
+      def mapReplacedTypeVarTo(to: Type) = new TypeMap:
+        override def apply(t: Type): Type =
+          if (t eq replacedTypeVar) && t.exists then to else mapOver(t)
+
+      var current = this
+
+      def removeParamFrom(ps: List[TypeParamRef]) =
+        ps.filterConserve(param ne _)
+
+      for lo <- lower(param) do
+        current = upperLens.map(this, current, lo, removeParamFrom)
+      for hi <- upper(param) do
+        current = lowerLens.map(this, current, hi, removeParamFrom)
+
+      def replaceParamIn(other: TypeParamRef) =
+        val oldEntry = current.entry(other)
+        val newEntry = current.ensureNonCyclic(other, oldEntry.substParam(param, replacement))
+        current = boundsLens.update(this, current, other, newEntry)
+        var oldDepEntry = oldEntry
+        var newDepEntry = newEntry
+        replacedTypeVar match
+          case tvar: TypeVar =>
+            if tvar.isInstantiated
+            then
+              // That's the case if replace is called from TypeVar's instantiateWith.
+              // Forget about instantiation for old dependencies.
+              oldDepEntry = mapReplacedTypeVarTo(param)(oldDepEntry)
+            else
+              // That's the case if replace is called from unify.
+              // Assume parameter has been replaced for new dependencies
+              // (the actual replacement is done below).
+              newDepEntry = mapReplacedTypeVarTo(replacement)(newDepEntry)
+          case _ =>
+        if oldDepEntry ne newDepEntry then
+          if current eq this then
+            // We can end up here if oldEntry eq newEntry, so posssibly no new constraint
+            // was created, but oldDepEntry ne newDepEntry. In that case we must make
+            // sure we have a new constraint before updating dependencies.
+            current = newConstraint()
+          current.adjustDeps(newDepEntry, oldDepEntry, other)
```

Comment:
Operations like `updateEntry` take `current` as an argument and check if it equals `this` to decide whether to copy or mutate the constraint, could `adjustDeps` do the same? The logic in replace is already very hard to follow so anything that can push complexity elsewhere would be welcome.

Reply:
I thought about it, but concluded that's not the same. `adjustDeps` is called to update internal data structures so that all internal invariants are satisfied. It does not create a new constraint, so there's generally no need to cache it. It's just in this specific corner case where it happens, and I tried to explain in the comment why.